### ISSUE TITLE
Socket has hardcoded "message" event

### ISSUE THREAD

**Issue by graforlock** [reasonml-community/bs-socket.io#9]
I believe Socket.on is not generic enough, as it stand now it is:

```ocaml
let on = (socket, func) =>
      _on(socket, "message", obj => func(Json.fromValidJson(obj)));
```

But looking at the docs [Socket.IO Documentation](https://socket.io/docs/), socket.io should support different types of events:

```javascript
io.on('connection', function(socket){
  socket.emit('request', /* */); // emit an event to the socket
  io.emit('broadcast', /* */); // emit an event to all connected sockets
  socket.on('reply', function(){ /* */ }); // listen to the event
});
```

**Comment by bsansouci** [comment]
Hey! Yup the design is to enable pattern matching, and socket.io’s design is closely tied to JS and it’s semantics. Why would you like to send stringly-typed messages?

One good reason could be to communicate with an existent backend/frontend. In which case I could imagine exposing some unsafe functions.

**Comment by graforlock** [comment]
Yes that is correct, if backend is in another language, the API should provide some flexibility for such case.
